src/main/java/org/spongepowered/api/service/permission/MemorySubjectData.java

```diff
@@ -51,20 +52,61 @@
  */
 public class MemorySubjectData implements SubjectData {
 
-    private final PermissionService service;
-    private final ConcurrentMap<Set<Context>, Map<String, String>> options = Maps.newConcurrentMap();
-    private final ConcurrentMap<Set<Context>, NodeTree> permissions = Maps.newConcurrentMap();
-    private final ConcurrentMap<Set<Context>, List<Map.Entry<String, String>>> parents = Maps.newConcurrentMap();
+    protected final Subject subject;
+    protected final PermissionService service;
+    protected final ConcurrentMap<Set<Context>, Map<String, String>> options = Maps.newConcurrentMap();
+    protected final ConcurrentMap<Set<Context>, NodeTree> permissions = Maps.newConcurrentMap();
+    protected final ConcurrentMap<Set<Context>, List<Map.Entry<String, String>>> parents = Maps.newConcurrentMap();
+
+    @Nullable
+    private Consumer<MemorySubjectData> updateListener = null;
 
     /**
      * Creates a new subject data instance, using the provided service to
      * request instances of permission subjects.
      *
      * @param service The service to request subjects from
      */
-    public MemorySubjectData(PermissionService service) {
-        checkNotNull(service, "service");
-        this.service = service;
+    public MemorySubjectData(Subject subject, PermissionService service) {
+        this.subject = checkNotNull(subject, "subject");
+        this.service = checkNotNull(service, "service");
+    }
+
+    /**
+     * Called each time the data in this {@link MemorySubjectData} is changed.
+     *
+     * <p>The default behaviour is to notify the change listener, if one is
+     * present.</p>
+     */
+    protected void onUpdate() {
+        if (this.updateListener != null) {
+            this.updateListener.accept(this);
+        }
```

Comment:
It gives the implementing plugin a chance to add any causes etc.

Also, since it is a requirement that the event is called async, a plugin instance is needed to schedule the event.